Deploy toolchain-member-operator on staging eaas cluster

amisstea opened the pull request.

amisstea commented:
@dperaza4dustbit @hugares please review. I'm not sure how to verify these changes before merging so I may be missing some things.

amisstea commented on argo-cd-apps/base/eaas/toolchain-member-operator/toolchain-member-operator.yaml:
```diff
@@ -0,0 +1,41 @@
+apiVersion: argoproj.io/v1alpha1
+kind: ApplicationSet
+metadata:
+  name: toolchain-member-operator-eaas
```
To avoid errors with `kustomize build` I gave this appset a unique name to avoid conflicts with `toolchain-member-operator` under the `../member/` dir.

```
> kustomize build argoc-cd-apps/overlays/development`
...
Error: accumulating resources: accumulation err='accumulating resources from '../../base/eaas': '/home/amisstea/src/infra-deployments/argo-cd-apps/base/eaas' must resolve to a file': recursed merging from path '/home/amisstea/src/infra-deployments/argo-cd-apps/base/eaas': may not add resource with an already registered id: ApplicationSet.v1alpha1.argoproj.io/toolchain-member-operator.argocd-staging
```

There may be a clever way to reorganize or reuse the appset with kustomize but I haven't found it yet. The problem is this appset has a different cluster generator than the existing `toolchain-member-operator` appset. Any suggestions would be appreciated.

hugares replied:
As discussed, the only way I found to reuse the same applicationSet for 2 type of clusters(was host and member in my case for backup), is to label the clusters with a new label and change the applicationSet to use this new label.

In the current case, we could label member and eaas clusters with a new label, e.g. `appstudio.redhat.com/toolchain-member`

amisstea commented:
/retest

amisstea commented:
@hugares please take another look. Here's some evidence of the what's changed with dev and staging by diffing the before and after of `kustomize build <overlay>`.

```
+ git checkout main
Switched to branch 'main'
Your branch is up to date with 'upstream/main'.
+ kustomize build argo-cd-apps/overlays/development
+ kustomize build argo-cd-apps/overlays/staging-downstream
+ kustomize build argo-cd-apps/overlays/konflux-public-staging
+ git checkout RHTAPWATCH-1147
Switched to branch 'RHTAPWATCH-1147'
+ kustomize build argo-cd-apps/overlays/development
+ kustomize build argo-cd-apps/overlays/staging-downstream
+ kustomize build argo-cd-apps/overlays/konflux-public-staging
+ diff -U 15 dev-original.yaml dev-new.yaml
--- dev-original.yaml   2024-07-18 11:46:06.662687312 -0400
+++ dev-new.yaml        2024-07-18 11:46:07.558692015 -0400
@@ -1,22 +1,23 @@
 apiVersion: v1
 kind: Secret
 metadata:
   labels:
     appstudio.redhat.com/host-cluster: "true"
     appstudio.redhat.com/internal-member-cluster: "true"
     appstudio.redhat.com/member-cluster: "true"
+    appstudio.redhat.com/toolchain-member: "true"
     argocd.argoproj.io/secret-type: cluster
   name: cluster-kubernetes.default.svc
   namespace: openshift-gitops
 stringData:
   config: '{"tlsClientConfig":{"insecure":false}}'
   name: in-cluster-local
   server: https://kubernetes.default.svc
 type: Opaque
 ---
 apiVersion: argoproj.io/v1alpha1
 kind: ApplicationSet
 metadata:
   name: application-api
   namespace: openshift-gitops
 spec:
+ diff -U 15 staging-ds-original.yaml staging-ds-new.yaml
--- staging-ds-original.yaml    2024-07-18 11:46:06.903688577 -0400
+++ staging-ds-new.yaml 2024-07-18 11:46:08.009694383 -0400
@@ -1796,31 +1796,31 @@
         syncOptions:
         - CreateNamespace=false
 ---
 apiVersion: argoproj.io/v1alpha1
 kind: ApplicationSet
 metadata:
   name: toolchain-member-operator
   namespace: argocd-staging
 spec:
   generators:
   - merge:
       generators:
       - clusters:
           selector:
             matchLabels:
-              appstudio.redhat.com/member-cluster: "true"
+              appstudio.redhat.com/toolchain-member: "true"
           values:
             clusterDir: ""
             environment: staging
             sourceRoot: components/sandbox/toolchain-member-operator
       - list:
           elements: []
       mergeKeys:
       - nameNormalized
   template:
     metadata:
       name: toolchain-member-operator-{{nameNormalized}}
     spec:
       destination:
         namespace: toolchain-member-operator
         server: '{{server}}'
+ diff -U 15 staging-public-original.yaml staging-public-new.yaml
--- staging-public-original.yaml        2024-07-18 11:46:07.205690162 -0400
+++ staging-public-new.yaml     2024-07-18 11:46:08.255695674 -0400
@@ -1829,31 +1829,31 @@
         syncOptions:
         - CreateNamespace=false
 ---
 apiVersion: argoproj.io/v1alpha1
 kind: ApplicationSet
 metadata:
   name: toolchain-member-operator
   namespace: konflux-public-staging
 spec:
   generators:
   - merge:
       generators:
       - clusters:
           selector:
             matchLabels:
-              appstudio.redhat.com/member-cluster: "true"
+              appstudio.redhat.com/toolchain-member: "true"
           values:
             clusterDir: ""
             environment: staging
             sourceRoot: components/sandbox/toolchain-member-operator
       - list:
           elements: []
       mergeKeys:
       - nameNormalized
   template:
     metadata:
       name: toolchain-member-operator-{{nameNormalized}}
     spec:
       destination:
         namespace: toolchain-member-operator
         server: '{{server}}'
```

hugares commented:
@amisstea I checked in ArgoCD instances to make sure the new label was applied to all member clusters and the only one without it is stg-m01: https://argocd-server-konflux-public-staging.apps.appsres05ue1.pwhm.p1.openshiftapps.com/settings/clusters/https%3A%2F%2Fapi.stone-stg-m01.7ayg.p1.openshiftapps.com%3A6443

I am not sure why yet, either we forgot this one in the app-interface MR or there is a deployment/refresh issue

amisstea commented:
@hugares the stg-m01 cluster has the label now.

hugares commented on argo-cd-apps/base/toolchain-member/kustomization.yaml:
```diff
@@ -3,4 +3,4 @@ kind: Kustomization
 components:
-  - ../../../../k-components/deploy-to-member-cluster-merge-generator
\ No newline at end of file
+  - ../../k-components/inject-infra-deployments-repo-details
```
is this not supposed to be `../../k-components/assign-toolchain-member-role-to-local-cluster`

amisstea replied:
This is intentional (see my other comment about it). That being said, I'm not sure if we need this `inject-infra-deployments-repo-details` k-component. It seems to be applied to some appsets but not others.

hugares commented on argo-cd-apps/base/toolchain-member/toolchain-member-operator.yaml:
```diff
@@ -9,6 +9,9 @@ spec:
           - clusters:
+              selector:
+                matchLabels:
+                  appstudio.redhat.com/toolchain-member: 'true'
```
should you use the `k-components/assign-toolchain-member-role-to-local-cluster` patch instead?

amisstea replied:
My understanding is that patch is meant for applying the label to the argocd local cluster secret for use in a single cluster dev env. I opted not to create a new k-component here, like backup, because it would only be applied to this appset.

hugares replied:
ok,  make sence, I reviewed too quickly

hugares reviewed (approved):
/lgtm
